[docs] Add Workflows REST API page (#46174)
expo/expo: docs/pages/eas/workflows/rest-api.mdx

---
title: Workflows REST API
sidebar_title: REST API

> description: Trigger an EAS Workflow and check its status from your own systems with the EAS REST API.

Reviewer: **[nit]** Glossary says "EAS Workflows" is always plural; "EAS Workflow" is the avoid form (`expo-docs-style-guide.md#glossary-and-product-names`). The lowercase singular noun reads more naturally here since you're talking about one workflow instance.

```suggestion
description: Trigger a workflow and check its status from your own systems with the EAS REST API.
```

Author: Thank you!

---

import { Dataflow03Icon } from '@expo/styleguide-icons/outline/Dataflow03Icon';

import { BoxLink } from '~/ui/components/BoxLink';

> The EAS REST API lets you trigger a workflow and check its status without going through the EAS CLI. Use it from your own CI system, a chat bot, an internal dashboard, or any service that can make HTTP requests.

Reviewer 2: nit: maybe it's just me, but the "Use it from…" sentence feels very much like an advertisement, like I read it and I'm like "sure i'll use it! [over eas-cli]" and i'm not sure we want to suggest people not to use eas-cli? idk… public-facing HTTP API is a common thing in our industry and yet we never exposed one so this feels like a bigger change

Reviewer: In that case, I'd suggest not mentioning EAS CLI here. Just the purpose of the REST API and what it does should be enough.

Author: Updated, thanks for the feedback!

All endpoints described below live under `https://api.expo.dev` and accept and return JSON.

## Authentication

Both endpoints require an EAS access token, sent as a bearer token in the `Authorization` header.

> For production integrations, create a [robot user](/accounts/programmatic-access/#robot-users-and-access-tokens) on the account that owns the project and give it a scoped role. The token then represents the bot, not a person, and you can revoke it without affecting any user. A [personal access token](/accounts/programmatic-access/#personal-access-tokens) also works for one-off scripts.

Reviewer: **[suggestion]** The middle sentence joins two distinct ideas with "and": who the token represents, and that it can be revoked without affecting users (`expo-docs-style-guide.md#voice-and-tone`). Splitting reads more like reference voice.

```suggestion
For production integrations, create a [robot user](/accounts/programmatic-access/#robot-users-and-access-tokens) on the account that owns the project and give it a scoped role. The token then represents the bot, not a person. You can revoke it without affecting any user. A [personal access token](/accounts/programmatic-access/#personal-access-tokens) also works for one-off scripts.
```

```text
Authorization: Bearer <EXPO_TOKEN>
Content-Type: application/json
```

## Trigger a workflow

```text
POST /v2/workflows/dispatch
```

Resolves the workflow file on the given Git ref, validates `inputs` against the `workflow_dispatch` schema declared in the workflow, and enqueues a new run.

### Request body

| Field      | Type          | Required | Description                                                                                                                  |
| ---------- | ------------- | -------- | ---------------------------------------------------------------------------------------------------------------------------- |

> | `appId`    | string (UUID) | Yes      | The EAS project ID. Find it on the project page or in **app.json** under `extra.eas.projectId`.                              |

Reviewer: ```suggestion
| `appId`    | string (UUID) | Yes      | The EAS project ID. Find it on the project page or in [app config](/workflow/configuration/) under `extra.eas.projectId`.                              |
```

See: https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md#referencing-appjsonappconfigjsonappconfigjsappconfigts

| `gitRef`   | string        | Yes      | A branch name (`main`), tag (`v1.0.0`), commit SHA, or fully qualified ref (`refs/heads/main`, `refs/tags/v1.0.0`).          |

> | `fileName` | string        | Yes      | The workflow file name only, such as `deploy.yml`. Do not include the **.eas/workflows/** prefix or any other path segments. |

Reviewer: ```suggestion
| `fileName` | string        | Yes      | The workflow file name only, such as **deploy.yml**. Do not include the **.eas/workflows/** prefix or any other path segments. |
```

See: https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md#file-names-directory-names-file-extensions-as-bold-text

| `inputs`   | object        | No       | Values for inputs declared under `on.workflow_dispatch.inputs` in the workflow file. Validated against the declared schema.  |

### Response

Returns `200 OK` with the new workflow run ID and a link to the run in the dashboard:

```json
{
  "data": {
    "id": "019d9d17-013a-7e05-89aa-4aa83ff68c32",
    "url": "https://expo.dev/accounts/acme/projects/example/workflows/019d9d17-013a-7e05-89aa-4aa83ff68c32"

>   }

Reviewer 2: theoretically there's no reason not to include the data we include in the regular GET here too

Author: The goal was to facilitate links to the dashboard for integrations that wish to dispatch with the new endpoint, rather than solely relying on the GET for it.  If you have strong opinions about making this more concise, please let me know 🙇‍♂️

}
```

### Example

```sh
curl -X POST "https://api.expo.dev/v2/workflows/dispatch" \
  -H "Authorization: Bearer $EXPO_TOKEN" \
  -H "Content-Type: application/json" \
  -d '{
    "appId": "a415eac6-231a-4b38-b481-3255a59f13b8",
    "gitRef": "main",
    "fileName": "deploy.yml",
    "inputs": { "environment": "production" }
  }'

> ```

Reviewer: **[design]** Shell commands shown to the reader must use the `Terminal` component, not a bare ` ```sh ` fence (`expo-docs-components.md#mandatory-usage`). You'll also need to add `import { Terminal } from '~/ui/components/Snippet';` near the top of the file.

```suggestion
<Terminal
  cmd={[
    `$ curl -X POST "https://api.expo.dev/v2/workflows/dispatch" \\`,
    `    -H "Authorization: Bearer $EXPO_TOKEN" \\`,
    `    -H "Content-Type: application/json" \\`,
    `    -d '{`,
    `      "appId": "a415eac6-231a-4b38-b481-3255a59f13b8",`,
    `      "gitRef": "main",`,
    `      "fileName": "deploy.yml",`,
    `      "inputs": { "environment": "production" }`,
    `    }'`,
  ]}
/>
```

| ----------------- | -------------------------------------------------------------------- |
| `new`             | The job is queued.                                                   |
| `in-progress`     | The job is running.                                                  |
| `action-required` | The job is paused and waiting on a manual action, such as approval.  |
| `pending-cancel`  | A cancellation has been requested but the job has not stopped yet.   |
| `success`         | The job succeeded. Terminal.                                         |
| `failure`         | The job failed. Terminal.                                            |
| `canceled`        | The job was canceled. Terminal.                                      |
| `skipped`         | The job was skipped because a required upstream job did not succeed. |

`outputs` contains any values the job set with `outputs:` in the workflow file. Secrets that were referenced when those values were written are replaced with placeholders before being returned. `buildId` and `submissionId` are present on jobs of `type: build` and `type: submission` and link to the underlying EAS Build or EAS Submit record.

### Example

> Trigger a run and poll until it reaches a terminal status:

Reviewer: Expanding to mention where to run the following bash script could be helpful like from a terminal or an actual script file.

Author: Updated, appreciate the input!

> ```sh

Reviewer: ````suggestion
```bash
````
